timecard/timecard.8
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2024 John Hay
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" Note: The date here should be updated whenever a non-trivial
.\" change is made to the manual page.
.Dd June 27, 2025
.Dt TIMECARD 8
.Os
.Sh NAME
.Nm timecard
.Nd "synchronise kernel time and provide ntpd(8) with time using an OCP TAP TimeCard"
.Sh SYNOPSIS
.Nm
.Op Fl a
.Op Fl B Ar cpuid
.Op Fl b
.Op Fl C Ar refv,mult
.Op Fl c
.Op Fl d Ar driftfile
.Op Fl f Ar /dev/timecardN
.Op Fl hk
.Op Fl l Ar logfile
.Op Fl r
.Op Fl s
.Op Fl T Ar period,min_shift,max_shift,stable
.Op Fl tv
.Sh DESCRIPTION
The
.Nm
does 3 main functions:
.Bl -tag -offset "xxxx"
.It Provide a shm(28) driver for ntpd.
.It Synchronise the kernel time to the
.Xr timecard 4
time.
.It Discipline or train the clock on board of the TimeCard through the
.Xr iic 4
Clock interface.
.El
.Pp
The options are as follows:
.Bl -tag -width "-d /dev/timecardN"
.It Fl a
Enable TimeCard clock aging.
.It Fl B Ar cpuid
Bind process to cpu.
.It Fl b
Enable BME monitoring, if it is compiled in.
.It Fl C Ar refv,mult
Enable clock temperature compensation. The refererence voltage (refv) and multiplier (mult) are floats.
.It Fl c
Enable TimeCard clock monitoring.
.It Fl d Ar driftfile
The file to periodically write drift values.
It is used to speed up TimeCard clock synchronization.
.It Fl f Ar /dev/timecardN
Specifiy which
.Pa /dev/timecard%d
to use.
If not specified,
.Pa /dev/timecard0
will be used.
.It Fl h
Enable a hot start.
Assume a clock that were already disciplined or a valid drift file.
.It Fl k
Enable kernel synchronisation.
.It Fl l Ar logfile
The logfile to use for per second logs. If not specified, none will be used.
.It Fl r
Read and display the time difference in nanoseconds between the TimeCard and Operating System (kernel) and then exit.
.It Fl s
Enable ntp shm shared memory interface.
.It Fl T Ar period,min_shift,max_shift,stable
Set the period (s), min, max shift (bits) and stable (ns) for training.
The defaults are 180,0,1,18.
.It Fl t
Enable clock training (disciplining).
.It Fl v
Be more verbose.
.El
.Pp
After initialization,
.Nm
will sit in a loop and once a second:
.Bl -tag -offset "xxxx" -compact
.It read the time and status from Xr timecard 4
.It check the card is still in sync and other sanity checks
.It update the ntp shm(28) interface
.It update the kernel time using Xr ntp_adjtime 2
.It read the Xr timecard 4 clock information
.It write the logfile
.It calculate Xr timecard 4 temperature compensation
.It calculate Xr timecard 4 clock disciplining / training
.It calculate Xr timecard 4 clock aging
.It update Xr timecard 4 clock pull to compensate for temperature changes
.It update Xr timecard 4 clock pull every training period
.It update Xr timecard 4 clock aging daily once settled
.It update the driftfile (hourly if it changed)
.El
.Pp
The logfile, has the following format:
.Pp
TC sec.nsec OS sec.nsec ClockStatusOffset ClockStatusDrift
ClockStatus TodSlaveStatus TodSlaveAntennaStatus TodSlaveUtcStatus PpsSlaveStatus State Valid
K TC-OS NOffs TSC-diff CPU
XO XOffs Xpwr IICuncorr IICerr
TR TOffs Tsec Offs sec Xadj TC ST Xaging Xpull
.Sh SIGNAL
If
.Nm
receives a SIGHUP signal,
it will close and reopen the logfile.
.Sh FILES
.Bl -tag -width "/dev/null" -compact
.It Pa /dev/timecard%d
.It Pa /dev/iic%d
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
The following is an example of a typical usage
of the
.Nm
command:
.Pp
.Dl "timecard -c -d driftfile.txt -k -l tc.log -s -t"
Option
.Fl k
can be left out if
.Xr ntpd 8
or similar will synchronise the kernel time.
.Dl "daemon -f -c timecard -c -d driftfile.txt -k -l tc.log -s -t"
.Sh DIAGNOSTICS
The command may fail for one of the following reasons:
.Bl -diag
.It "Could not find IIC devices."
Self explanatory.
.It "Could not open all devices."
Self explanatory.
.It "Error initializing EEPROM %d"
The error is not fatal.
.El
.Sh SEE ALSO
.Xr timecard 4 ,
.Xr iic 4
.Lk https://github.com/opencomputeproject/Time-Appliance-Project/tree/master/Time-Card
.Sh HISTORY
.Nm
is part of the
.Xr timecard 4
suite.
.Sh BUGS
The only clock that is supported is the SiTime SiT5721.


timecard/timecard.sh
#!/bin/sh

# PROVIDE: timecard
# REQUIRE: syslogd
# BEFORE: SERVERS ntpd
# KEYWORD:

#
# Add the following lines to /etc/rc.conf to enable timecard:
#
# timecard_enable (bool):	Set to "NO" by default.
#				Set it to "YES" to enable timecard.
# timecard_flags (str):	Set to "" by default.
#				See timecard(8) options for details.
# timecard_logfile (str):
# timecard_driftfile (str):
# timecard_hints (str):		See timecard_default_hints below.
# timecard_modules (str):	See timecard_default_modules below.
# timecard_pps_symlink (str):	Source_dev target_dev, eg.
#				"/dev/timecard0 /dev/pps0"
# timecard_servo (str):		See options for tc-adj-servo(8).
# timecard_sync_wait (num):	Seconds to wait for card to sync.
#				Default is 30 seconds.

. /etc/rc.subr

name="timecard"
desc="Synchronize kernel time to the TimeCard and provide shm(28) driver for ntpd"
rcvar="timecard_enable"

timecard_default_modules="
	axi_iic:timecard/axi_iic
	iic:iicbus/iic
	axi_spi_timecard:timecard/spi
	spigen:spibus/spigen
	uart_timecard:timecard/uart
	timecard:pci/timecard
"
timecard_default_hints="
	hint.spigen.0.at=spibus0
	hint.spigen.0.clock=1000000
	hint.spigen.0.cs=0
	hint.spigen.0.mode=0
"

pidfile="/var/run/${name}.pid"
procname="/usr/local/sbin/${name}"
command=/usr/sbin/daemon
command_args="-f -c -T timecard -p ${pidfile} ${procname}"
tc_adj_servo="/usr/local/sbin/tc-adj-servo"

start_cmd="timecard_start"
start_precmd="timecard_prestart"

load_rc_config $name

# Set defaults
: ${timecard_enable:="NO"}
: ${timecard_hints:=${timecard_default_hints}}
: ${timecard_insync_ns:=100000000}
: ${timecard_sync_wait:=30}
: ${required_modules:=${timecard_modules}}
: ${required_modules:=${timecard_default_modules}}

timecard_pre_wait()
{
	local to=${timecard_sync_wait}
	local insync
	local utcvalid
	local fixok

	while [ ${to} -gt 0 ]; do
		insync=`sysctl -n dev.timecard.0.status.clk_in_sync`
		utcvalid=`sysctl -n dev.timecard.0.status.tod_utc_valid`
		fixok=`sysctl -n dev.timecard.0.status.gnss_fix_ok`
		if [ ${insync} -eq 1 -a ${utcvalid} -eq 1 -a ${fixok} -eq 1 ]; then
			break;
		fi
		sleep 1
		to=$((${to}-1))
	done
}

timecard_post_wait()
{
	local to=${timecard_sync_wait}
	local rdline
	local diff

	while [ ${to} -gt 0 ]; do
		rdline=`${procname} -r`
		diff=${rdline##*DIFF }
		diff=${diff%% TC*}
		if [ ${diff} -lt ${timecard_insync_ns} ]; then
			break
		fi
		sleep 1
		to=$((${to}-1))
	done
}

timecard_prestart()
{
	local _h
	for _h in ${timecard_hints}; do
		kenv -q ${_h}
	done
}

timecard_start()
{
	if [ ! -c "/dev/timecard0" ]; then
		echo "timecard driver did not load!"
		return 6
	fi
	if [ -n "${timecard_driftfile}" ]; then
		#command_args="${command_args} -d ${timecard_driftfile}"
		timecard_flags="${timecard_flags} -d ${timecard_driftfile}"
	fi
	if [ -n "${timecard_logfile}" ]; then
		timecard_flags="${timecard_flags} -l ${timecard_logfile}"
	fi
	if [ -n "${timecard_servo}" ]; then
		${tc_adj_servo} ${timecard_servo}
	fi
	if [ -n "${timecard_pps_symlink}" ]; then
		ln -f -s ${timecard_pps_symlink}
	fi
	timecard_pre_wait
	/sbin/sysctl kern.timecounter.hardware="TimeCard"
	$command $command_args $timecard_flags
	timecard_post_wait
}

run_rc_command "$1"

